src/dcd/common/constants.d

```diff
@@ -20,187 +20,986 @@ module dcd.common.constants;
+{
+	pragma(inline, true);
+	++x;
+	pragma(inline, false); // supercedes the others
+	return x + 3;
+}
+---`),
+	ConstantCompletion("lib", `Inserts a directive in the object file to link in the library
+specified by the $(ASSIGNEXPRESSION).
+The $(ASSIGNEXPRESSION)s must be a string literal:
+-----------------
+pragma(lib, "foo.lib");
+-----------------`),
+	ConstantCompletion("mangle", `Overrides the default mangling for a symbol. It's only effective
+when the symbol is a function declaration or a variable declaration.
+For example this allows linking to a symbol which is a D keyword, which would normally
+be disallowed as a symbol name:
+-----------------
+pragma(mangle, "body")
+extern(C) void body_func();
+-----------------`),
+	ConstantCompletion("msg", `Constructs a message from the arguments and prints to the standard error stream while compiling:
+-----------------
+pragma(msg, "compiling...", 1, 1.0);
+-----------------`),
+	ConstantCompletion("startaddress", `Puts a directive into the object file saying that the
+function specified in the first argument will be the
+start address for the program:
+-----------------
+void foo() { ... }
+pragma(startaddress, foo);
+-----------------
+This is not normally used for application level programming,
+but is for specialized systems work.
+For applications code, the start address is taken care of
+by the runtime library.`)
 ];
 
 /**
  * Linkage types
  */
-immutable string[] linkages = [
-	"C",
-	"C++",
-	"D",
-	"Objective-C",
-	"Pascal",
-	"System",
-	"Windows"
+immutable ConstantCompletion[] linkages = [
+	// https://dlang.org/spec/attribute.html#linkage
+	ConstantCompletion("C", "Enforces C calling conventions for the function, no mangling"),
+	ConstantCompletion("C++", "offers limited compatibility with C++"),
+	ConstantCompletion("D", "Default D mangling and calling conventions"),
+	ConstantCompletion("Objective-C",
+		"Objective-C offers limited compatibility with Objective-C, see the Interfacing to Objective-C documentation for more information"),
+	ConstantCompletion("Pascal"),
+	ConstantCompletion("System", "`Windows` on Windows platforms, `C` on other platforms."),
+	ConstantCompletion("Windows", "Enforces Win32/`__stdcall` conventions for the function")
 ];
 
+private immutable string isLazyDoc = `$(P Takes one argument. If that argument is a declaration,
+	$(D true) is returned if it is $(D_KEYWORD ref), $(D_KEYWORD out),
+	or $(D_KEYWORD lazy), otherwise $(D false).
+)
+
+---
+void fooref(ref int x)
+{
+	static assert(__traits(isRef, x));
+	static assert(!__traits(isOut, x));
+	static assert(!__traits(isLazy, x));
+}
+
+void fooout(out int x)
+{
+	static assert(!__traits(isRef, x));
+	static assert(__traits(isOut, x));
+	static assert(!__traits(isLazy, x));
+}
+
+void foolazy(lazy int x)
+{
+	static assert(!__traits(isRef, x));
+	static assert(!__traits(isOut, x));
+	static assert(__traits(isLazy, x));
+}
+---`;
+
 /**
  * Traits arguments
  */
```

Comment:
This looks like it'll be a pain to keep synchronized with the dlang.org documentation.

Reply:
hm how about doing a .txt file where we just copy paste the content and then with some CTFE it extracts the strings?

Reply:
Wouldn't that leave us with the obligation to maintain the documentation AND the new code that extracts it with CTFE? My concern here is that this documentation is now part of the DCD repository where it wasn't before, not that it's in D code instead of something else.

Reply:
hm imo the documentation really made it really a lot more noob friendly, so I think it should be included. Of course we could instead make a HTTP API but that would require an active internet connection and introduces delays etc. Another possibility would be providing a dub package for documentation, but that one should then have all language documentation and I think maintaining that is a bigger problem than just 3 sections. Actually I quite like the idea of having a dub package for documentation but having it come from a third party instead of making sure there are guidelines to editing the pages so they can be auto generated makes it a lot more unappealing.